File: agents/github-action-agent/src/tonic_agent/prompt_engineering.py
"""Prompt templates (Tonic semantics; strings from shared prompt bundle)."""

from __future__ import annotations

from enum import Enum

from . import env_config
from .models import ConflictFile, ConflictRegion
from . import prompt_bundle


class PromptTemplate(Enum):
    Default = "default"
    Enhanced = "enhanced"
    ContextAware = "context-aware"


def _bundle_key(t: PromptTemplate) -> str:
    match t:
        case PromptTemplate.Default:
            return "default"
        case PromptTemplate.Enhanced:
            return "enhanced"
        case PromptTemplate.ContextAware:
            return "context_aware"


class PromptGenerator:
    def __init__(self, template: PromptTemplate) -> None:
        self.template = template
        self._bk = _bundle_key(template)

    def generate_system_prompt(self) -> str:
        custom = env_config.get_system_prompt_override()
        if custom is not None:
            return custom
        return prompt_bundle.system_prompt_body(self._bk)

    def generate_conflict_prompt(
        self,
        conflict_file: ConflictFile,
        conflict: ConflictRegion,
        *,
        expected_resolved_line_count: int | None = None,
    ) -> str:
        match self.template:
            case PromptTemplate.Default:
                kind_suffix = f" ({conflict.conflict_kind})" if conflict.conflict_kind else ""
                message = prompt_bundle.format_conflict_user(
                    self._bk,
                    path=conflict_file.path,
                    start_line=str(conflict.start_line),
                    end_line=str(conflict.end_line),
                    kind_suffix=kind_suffix,
                    left_label=conflict_file.left_label,
                    right_label=conflict_file.right_label,
                    left_content=conflict.left_content,
                    right_content=conflict.right_content,
                )
                return _append_line_count_guidance(message, expected_resolved_line_count)
            case PromptTemplate.Enhanced:
                kind = conflict.conflict_kind or "unspecified"
                ft = determine_file_type(conflict_file.path)
                message = prompt_bundle.format_conflict_user(
                    self._bk,
                    path=conflict_file.path,
                    start_line=str(conflict.start_line),
                    end_line=str(conflict.end_line),
                    kind=kind,
                    file_type=ft,
                    left_label=conflict_file.left_label,
                    right_label=conflict_file.right_label,
                    left_content=conflict.left_content,
                    right_content=conflict.right_content,
                )
                return _append_line_count_guidance(message, expected_resolved_line_count)
            case PromptTemplate.ContextAware:
                base_section = (
                    f"BASE VERSION (common ancestor):\n```\n{conflict.base_content}```\n\n"
                    if conflict.base_content.strip()
                    else ""
                )
                surrounding = extract_surrounding_context(conflict_file, conflict)
                ctx_section = (
                    f"SURROUNDING CONTEXT:\n```\n{surrounding}```\n\n" if surrounding.strip() else ""
                )
                ft = determine_file_type(conflict_file.path)
                kind = conflict.conflict_kind or "unspecified"
                message = prompt_bundle.format_conflict_user(
                    self._bk,
                    path=conflict_file.path,
                    start_line=str(conflict.start_line),
                    end_line=str(conflict.end_line),
                    kind=kind,
                    file_type=ft,
                    base_section=base_section,
                    ctx_section=ctx_section,
                    left_label=conflict_file.left_label,
                    right_label=conflict_file.right_label,
                    left_content=conflict.left_content,
                    right_content=conflict.right_content,
                )
                return _append_line_count_guidance(message, expected_resolved_line_count)

    def generate_file_prompt(self, conflict_file: ConflictFile) -> str:
        match self.template:
            case PromptTemplate.Default:
                parts: list[str] = []
                for i, c in enumerate(conflict_file.conflicts):
                    parts.append(
                        f"CONFLICT {i + 1}:\nBetween lines {c.start_line} and {c.end_line}\n"
                        f"LEFT:\n```\n{c.left_content}```\n"
                        f"RIGHT:\n```\n{c.right_content}```\n\n"
                    )
                body = "".join(parts)
                return prompt_bundle.format_file_user(
                    self._bk,
                    path=conflict_file.path,
                    conflict_count=str(len(conflict_file.conflicts)),
                    conflicts_body=body,
                )
            case PromptTemplate.Enhanced:
                ft = determine_file_type(conflict_file.path)
                parts = []
                for i, c in enumerate(conflict_file.conflicts):
                    parts.append(
                        f"CONFLICT {i + 1}:\n"
                        f"- Location: Lines {c.start_line} to {c.end_line}\n"
                        f"- Kind: {c.conflict_kind or 'unspecified'}\n"
                        f"LEFT:\n```\n{c.left_content}```\n"
                        f"RIGHT:\n```\n{c.right_content}```\n\n"
                    )
                return prompt_bundle.format_file_user(
                    self._bk,
                    path=conflict_file.path,
                    file_type=ft,
                    conflict_count=str(len(conflict_file.conflicts)),
                    conflicts_body="".join(parts),
                )
            case PromptTemplate.ContextAware:
                ft = determine_file_type(conflict_file.path)
                parts = []
                for i, c in enumerate(conflict_file.conflicts):
                    base_sec = f"BASE:\n```\n{c.base_content}```\n" if c.base_content.strip() else ""
                    parts.append(
                        f"CONFLICT {i + 1}:\n"
                        f"- Lines {c.start_line} to {c.end_line}\n"
                        f"{base_sec}"
                        f"LEFT:\n```\n{c.left_content}```\n"
                        f"RIGHT:\n```\n{c.right_content}```\n\n"
                    )
                return prompt_bundle.format_file_user(
                    self._bk,
                    path=conflict_file.path,
                    file_type=ft,
                    conflict_count=str(len(conflict_file.conflicts)),
                    conflicts_body="".join(parts),
                )


def determine_file_type(path: str) -> str:
    if "." not in path:
        return "Unknown file type"
    extension = path.rsplit(".", 1)[-1].lower()
    match extension:
        case "rs":
            return "Rust"
        case "js":
            return "JavaScript"
        case "ts":
            return "TypeScript"
        case "py":
            return "Python"
        case "md":
            return "Markdown"
        case "json":
            return "JSON"
        case "yml" | "yaml":
            return "YAML"
        case "toml":
            return "TOML"
        case _:
            return f"File with .{extension} extension"


def extract_surrounding_context(conflict_file: ConflictFile, conflict: ConflictRegion) -> str:
    if not conflict_file.content:
        return ""
    lines = conflict_file.content.splitlines()
    start = max(0, conflict.start_line - 3)
    end = min(len(lines), conflict.end_line + 2)
    return "\n".join(lines[start:end])


def build_expected_resolved_line_count_guidance(expected_resolved_line_count: int | None) -> str:
    if expected_resolved_line_count is None or expected_resolved_line_count <= 0:
        return ""
    noun = "line" if expected_resolved_line_count == 1 else "lines"
    return (
        "Important output constraint: return exactly "
        f"{expected_resolved_line_count} {noun} in resolved_lines so the "
        "suggestion matches the current head-side span length."
    )


def _append_line_count_guidance(message: str, expected_resolved_line_count: int | None) -> str:
    guidance = build_expected_resolved_line_count_guidance(expected_resolved_line_count)
    if not guidance:
        return message
    return f"{message}\n\n{guidance}"


def prompt_template_from_env() -> PromptTemplate:
    raw = (env_config.get_prompt_template_name() or "enhanced").lower().strip()
    if raw == "default":
        return PromptTemplate.Default
    if raw in ("context-aware", "context_aware", "contextaware"):
        return PromptTemplate.ContextAware
    return PromptTemplate.Enhanced

File: agents/github-action-agent-node/src/promptEngineering.ts
import type { ConflictFile, ConflictRegion } from "@mergetonic/core";

import bundleJson from "./data/aiPrompts.v1.json";

type BundleV1 = {
  schema_version: number;
  github_json_response_suffix: string;
  system_prompts: Record<"default" | "enhanced" | "context_aware", string>;
  conflict_user: Record<"default" | "enhanced" | "context_aware", string>;
  file_user: Record<"default" | "enhanced" | "context_aware", string>;
};

const bundle = bundleJson as BundleV1;

export type PromptTemplateEnv = "default" | "enhanced" | "context-aware";

function bundleKey(t: PromptTemplateEnv): keyof BundleV1["system_prompts"] {
  if (t === "context-aware") {
    return "context_aware";
  }
  return t;
}

export function promptTemplateFromEnv(): PromptTemplateEnv {
  const raw = (process.env.TONIC_AGENT_PROMPT_TEMPLATE ?? "enhanced")
    .toLowerCase()
    .trim();
  if (raw === "default") {
    return "default";
  }
  if (raw === "context-aware" || raw === "context_aware" || raw === "contextaware") {
    return "context-aware";
  }
  return "enhanced";
}

function formatTpl(tpl: string, vars: Record<string, string>): string {
  return tpl.replace(/\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g, (_, key: string) =>
    Object.prototype.hasOwnProperty.call(vars, key) ? vars[key]! : `{${key}}`,
  );
}

export function githubJsonResponseSuffix(): string {
  return bundle.github_json_response_suffix;
}

export function buildSystemPromptBody(template: PromptTemplateEnv): string {
  return bundle.system_prompts[bundleKey(template)];
}

export function determineFileType(filePath: string): string {
  if (!filePath.includes(".")) {
    return "Unknown file type";
  }
  const extension = filePath.split(".").pop()!.toLowerCase();
  switch (extension) {
    case "rs":
      return "Rust";
    case "js":
      return "JavaScript";
    case "ts":
      return "TypeScript";
    case "py":
      return "Python";
    case "md":
      return "Markdown";
    case "json":
      return "JSON";
    case "yml":
    case "yaml":
      return "YAML";
    case "toml":
      return "TOML";
    default:
      return `File with .${extension} extension`;
  }
}

export function extractSurroundingContext(conflictFile: ConflictFile, conflict: ConflictRegion): string {
  if (!conflictFile.content) {
    return "";
  }
  const lines = conflictFile.content.split(/\r?\n/);
  const start = Math.max(0, conflict.startLine - 3);
  const end = Math.min(lines.length, conflict.endLine + 2);
  return lines.slice(start, end).join("\n");
}

export function buildConflictUserMessage(
  cf: ConflictFile,
  conflict: ConflictRegion,
  template: PromptTemplateEnv,
): string {
  const leftLabel = cf.leftLabel ?? "left";
  const rightLabel = cf.rightLabel ?? "right";
  const bk = bundleKey(template);
  const tpl = bundle.conflict_user[bk];

  if (template === "default") {
    const kindSuffix = conflict.conflictKind ? ` (${conflict.conflictKind})` : "";
    return formatTpl(tpl, {
      path: cf.path,
      start_line: String(conflict.startLine),
      end_line: String(conflict.endLine),
      kind_suffix: kindSuffix,
      left_label: leftLabel,
      right_label: rightLabel,
      left_content: conflict.leftContent ?? "",
      right_content: conflict.rightContent ?? "",
    });
  }

  if (template === "enhanced") {
    const kind = conflict.conflictKind || "unspecified";
    const ft = determineFileType(cf.path);
    return formatTpl(tpl, {
      path: cf.path,
      start_line: String(conflict.startLine),
      end_line: String(conflict.endLine),
      kind,
      file_type: ft,
      left_label: leftLabel,
      right_label: rightLabel,
      left_content: conflict.leftContent ?? "",
      right_content: conflict.rightContent ?? "",
    });
  }

  const baseSection =
    conflict.baseContent?.trim() ?
      `BASE VERSION (common ancestor):\n\`\`\`\n${conflict.baseContent}\`\`\`\n\n`
    : "";
  const surrounding = extractSurroundingContext(cf, conflict);
  const ctxSection = surrounding.trim()
    ? `SURROUNDING CONTEXT:\n\`\`\`\n${surrounding}\`\`\`\n\n`
    : "";
  const kind = conflict.conflictKind || "unspecified";
  const ft = determineFileType(cf.path);
  return formatTpl(tpl, {
    path: cf.path,
    start_line: String(conflict.startLine),
    end_line: String(conflict.endLine),
    kind,
    file_type: ft,
    base_section: baseSection,
    ctx_section: ctxSection,
    left_label: leftLabel,
    right_label: rightLabel,
    left_content: conflict.leftContent ?? "",
    right_content: conflict.rightContent ?? "",
  });
}

export function buildFileUserMessage(cf: ConflictFile, template: PromptTemplateEnv): string {
  const bk = bundleKey(template);
  const tpl = bundle.file_user[bk];
  const n = cf.conflicts.length;

  if (template === "default") {
    const parts: string[] = [];
    for (let i = 0; i < cf.conflicts.length; i++) {
      const c = cf.conflicts[i]!;
      parts.push(
        `CONFLICT ${i + 1}:\nBetween lines ${c.startLine} and ${c.endLine}\n` +
          `LEFT:\n\`\`\`\n${c.leftContent ?? ""}\`\`\`\n` +
          `RIGHT:\n\`\`\`\n${c.rightContent ?? ""}\`\`\`\n\n`,
      );
    }
    return formatTpl(tpl, {
      path: cf.path,
      conflict_count: String(n),
      conflicts_body: parts.join(""),
    });
  }

  if (template === "enhanced") {
    const ft = determineFileType(cf.path);
    const parts: string[] = [];
    for (let i = 0; i < cf.conflicts.length; i++) {
      const c = cf.conflicts[i]!;
      parts.push(
        `CONFLICT ${i + 1}:\n` +
          `- Location: Lines ${c.startLine} to ${c.endLine}\n` +
          `- Kind: ${c.conflictKind || "unspecified"}\n` +
          `LEFT:\n\`\`\`\n${c.leftContent ?? ""}\`\`\`\n` +
          `RIGHT:\n\`\`\`\n${c.rightContent ?? ""}\`\`\`\n\n`,
      );
    }
    return formatTpl(tpl, {
      path: cf.path,
      file_type: ft,
      conflict_count: String(n),
      conflicts_body: parts.join(""),
    });
  }

  const ft = determineFileType(cf.path);
  const parts: string[] = [];
  for (let i = 0; i < cf.conflicts.length; i++) {
    const c = cf.conflicts[i]!;
    const baseSec = c.baseContent?.trim() ? `BASE:\n\`\`\`\n${c.baseContent}\`\`\`\n` : "";
    parts.push(
      `CONFLICT ${i + 1}:\n` +
        `- Lines ${c.startLine} to ${c.endLine}\n` +
        `${baseSec}` +
        `LEFT:\n\`\`\`\n${c.leftContent ?? ""}\`\`\`\n` +
        `RIGHT:\n\`\`\`\n${c.rightContent ?? ""}\`\`\`\n\n`,
    );
  }
  return formatTpl(tpl, {
    path: cf.path,
    file_type: ft,
    conflict_count: String(n),
    conflicts_body: parts.join(""),
  });
}
